File: rust/2024/src/bin/day04.rs
#[derive(Clone)]
struct Grid {
    letters: Vec<char>,
    size: usize,
}

impl From<&str> for Grid {
    fn from(input: &str) -> Self {
        let letters = input.chars().filter(|&c| c != '\n').collect::<Vec<_>>();
        let size = letters.len().isqrt();

        Self { letters, size }
    }
}

type Index<'a> = Box<dyn Iterator<Item = (usize, usize, usize, usize)> + 'a>;

impl Grid {
    fn iter(&self) -> GridIterator<'_> {
        GridIterator {
            grid: self,
            index: self.horizontal_index(),
            method: GridIterationMethod::Horizontal,
        }
    }

    fn cross_iter(&self) -> GridIterator<'_> {
        GridIterator {
            grid: self,
            index: self.cross_index(),
            method: GridIterationMethod::Cross,
        }
    }

    fn horizontal_index(&self) -> Index<'_> {
        Box::new((0..self.size - 3).flat_map(move |x| {
            (0..self.size).filter_map(move |y| {
                if matches!(self.letters[x + y * self.size], 'X' | 'S') {
                    Some((
                        x + y * self.size,
                        x + 1 + y * self.size,
                        x + 2 + y * self.size,
                        x + 3 + y * self.size,
                    ))
                } else {
                    None
                }
            })
        }))
    }

    fn vertical_index(&self) -> Index<'_> {
        Box::new((0..self.size).flat_map(move |x| {
            (0..self.size - 3).filter_map(move |y| {
                if matches!(self.letters[x + y * self.size], 'X' | 'S') {
                    Some((
                        x + y * self.size,
                        x + (y + 1) * self.size,
                        x + (y + 2) * self.size,
                        x + (y + 3) * self.size,
                    ))
                } else {
                    None
                }
            })
        }))
    }

    fn diagonal_up_index(&self) -> Index<'_> {
        Box::new((0..self.size - 3).flat_map(move |x| {
            (3..self.size).filter_map(move |y| {
                if matches!(self.letters[x + y * self.size], 'X' | 'S') {
                    Some((
                        x + y * self.size,
                        x + 1 + (y - 1) * self.size,
                        x + 2 + (y - 2) * self.size,
                        x + 3 + (y - 3) * self.size,
                    ))
                } else {
                    None
                }
            })
        }))
    }

    fn diagonal_down_index(&self) -> Index<'_> {
        Box::new((0..self.size - 3).flat_map(move |x| {
            (0..self.size - 3).filter_map(move |y| {
                if matches!(self.letters[x + y * self.size], 'X' | 'S') {
                    Some((
                        x + y * self.size,
                        x + 1 + (y + 1) * self.size,
                        x + 2 + (y + 2) * self.size,
                        x + 3 + (y + 3) * self.size,
                    ))
                } else {
                    None
                }
            })
        }))
    }

    fn cross_index(&self) -> Index<'_> {
        Box::new((0..self.size - 2).flat_map(move |x| {
            (0..self.size - 2).filter_map(move |y| {
                if self.letters[x + 1 + (y + 1) * self.size] == 'A' {
                    Some((
                        x + y * self.size,
                        x + 2 + y * self.size,
                        x + 2 + (y + 2) * self.size,
                        x + (y + 2) * self.size,
                    ))
                } else {
                    None
                }
            })
        }))
    }
}

enum GridIterationMethod {
    Horizontal,
    Vertical,
    DiagonalUp,
    DiagonalDown,
    Cross,
}

struct GridIterator<'a> {
    grid: &'a Grid,
    index: Index<'a>,
    method: GridIterationMethod,
}

impl Iterator for GridIterator<'_> {
    type Item = (char, char, char, char);

    fn next(&mut self) -> Option<Self::Item> {
        let mut next = self.index.next();

        if next.is_none() {
            match self.method {
                GridIterationMethod::Horizontal => {
                    self.method = GridIterationMethod::Vertical;
                    self.index = self.grid.vertical_index();
                }
                GridIterationMethod::Vertical => {
                    self.method = GridIterationMethod::DiagonalUp;
                    self.index = self.grid.diagonal_up_index();
                }
                GridIterationMethod::DiagonalUp => {
                    self.method = GridIterationMethod::DiagonalDown;
                    self.index = self.grid.diagonal_down_index();
                }
                GridIterationMethod::DiagonalDown | GridIterationMethod::Cross => {
                    return None;
                }
            }

            next = self.index.next();
        }

        if let Some((a, b, c, d)) = next {
            Some((
                self.grid.letters[a],
                self.grid.letters[b],
                self.grid.letters[c],
                self.grid.letters[d],
            ))
        } else {
            None
        }
    }
}

#[derive(Clone)]
struct Solution {
    grid: Grid,
}

impl Solver for Solution {
    fn new(input: &str) -> Anyhow<Self> {
        Ok(Self {
            grid: Grid::from(input),
        })
    }

    fn part1(&mut self) -> Anyhow<impl fmt::Display> {
        Ok(self
            .grid
            .iter()
            .filter(|tup| matches!(tup, ('X', 'M', 'A', 'S') | ('S', 'A', 'M', 'X')))
            .count())
    }

    fn part2(&mut self) -> Anyhow<impl fmt::Display> {
        Ok(self
            .grid
            .cross_iter()
            .filter(|tup| {
                matches!(
                    tup,
                    ('M', 'M', 'S', 'S')
                        | ('M', 'S', 'S', 'M')
                        | ('S', 'S', 'M', 'M')
                        | ('S', 'M', 'M', 'S')
                )
            })
            .count())
    }
}

aoc::solution!();

#[cfg(test)]
mod test {
    use super::{Solution, Solver};

    const INPUT: &str = r"MMMSXXMASM
MSAMXMSMSA
AMXSXMAAMM
MSAMASMSMX
XMASAMXAMM
XXAMMXXAMA
SMSMSASXSS
SAXAMASAAA
MAMMMXMMMM
MXMXAXMASX";

    #[test]
    fn test_part1() {
        let mut solution = Solution::new(INPUT).unwrap();
        let answer = solution.part1().unwrap().to_string();
        assert_eq!(answer, "18");
    }

    #[test]
    fn test_part2() {
        let mut solution = Solution::new(INPUT).unwrap();
        let answer = solution.part2().unwrap().to_string();
        assert_eq!(answer, "9");
    }
}

File: python/2024/day04.py
from __future__ import annotations

from enum import Enum, auto
from math import sqrt
from typing import TYPE_CHECKING, Self

if TYPE_CHECKING:
    from collections.abc import Generator


class GridIterationMethod(Enum):
    Horizontal = auto()
    Vertical = auto()
    DiagonalUp = auto()
    DiagonalDown = auto()
    Cross = auto()


class Grid:
    def __init__(self, input: str) -> None:
        self.letters = "".join(input.splitlines())
        self.size = int(sqrt(len(self.letters)))
        self.method = GridIterationMethod.Horizontal
        self.index = self.horizontal_index()

    def cross(self) -> Self:
        self.method = GridIterationMethod.Cross
        self.index = self.cross_index()
        return self

    def horizontal_index(self) -> Generator[tuple[int, int, int, int] | None]:
        for x in range(0, self.size - 3):
            for y in range(0, self.size):
                if self.letters[x + y * self.size] in ("X", "S"):
                    yield (
                        x + y * self.size,
                        x + 1 + y * self.size,
                        x + 2 + y * self.size,
                        x + 3 + y * self.size,
                    )
        yield None

    def vertical_index(self) -> Generator[tuple[int, int, int, int] | None]:
        for x in range(0, self.size):
            for y in range(0, self.size - 3):
                if self.letters[x + y * self.size] in ("X", "S"):
                    yield (
                        x + y * self.size,
                        x + (y + 1) * self.size,
                        x + (y + 2) * self.size,
                        x + (y + 3) * self.size,
                    )
        yield None

    def diagonal_up_index(self) -> Generator[tuple[int, int, int, int] | None]:
        for x in range(0, self.size - 3):
            for y in range(3, self.size):
                if self.letters[x + y * self.size] in ("X", "S"):
                    yield (
                        x + y * self.size,
                        x + 1 + (y - 1) * self.size,
                        x + 2 + (y - 2) * self.size,
                        x + 3 + (y - 3) * self.size,
                    )
        yield None

    def diagonal_down_index(self) -> Generator[tuple[int, int, int, int] | None]:
        for x in range(0, self.size - 3):
            for y in range(0, self.size - 3):
                if self.letters[x + y * self.size] in ("X", "S"):
                    yield (
                        x + y * self.size,
                        x + 1 + (y + 1) * self.size,
                        x + 2 + (y + 2) * self.size,
                        x + 3 + (y + 3) * self.size,
                    )
        yield None

    def cross_index(self) -> Generator[tuple[int, int, int, int] | None]:
        for x in range(0, self.size - 2):
            for y in range(0, self.size - 2):
                if self.letters[x + 1 + (y + 1) * self.size] == "A":
                    yield (
                        x + y * self.size,
                        x + 2 + y * self.size,
                        x + 2 + (y + 2) * self.size,
                        x + (y + 2) * self.size,
                    )
        yield None

    def __iter__(self) -> Self:
        return self

    def __next__(self) -> tuple[str, str, str, str]:
        index = next(self.index)

        if index is None:
            match self.method:
                case GridIterationMethod.Horizontal:
                    self.method = GridIterationMethod.Vertical
                    self.index = self.vertical_index()
                case GridIterationMethod.Vertical:
                    self.method = GridIterationMethod.DiagonalUp
                    self.index = self.diagonal_up_index()
                case GridIterationMethod.DiagonalUp:
                    self.method = GridIterationMethod.DiagonalDown
                    self.index = self.diagonal_down_index()
                case GridIterationMethod.DiagonalDown | GridIterationMethod.Cross:
                    raise StopIteration

            index = next(self.index)

        match index:
            case (a, b, c, d):
                return (
                    self.letters[a],
                    self.letters[b],
                    self.letters[c],
                    self.letters[d],
                )
            case _:
                raise StopIteration


def part1(input: str) -> int:
    def f(tup: tuple[str, str, str, str]) -> bool:
        match tup:
            case ("X", "M", "A", "S") | ("S", "A", "M", "X"):
                return True
            case _:
                return False

    return len(list(filter(f, Grid(input))))


def part2(input: str) -> int:
    def f(tup: tuple[str, str, str, str]) -> bool:
        match tup:
            case (
                ("M", "M", "S", "S")
                | ("M", "S", "S", "M")
                | ("S", "S", "M", "M")
                | ("S", "M", "M", "S")
            ):
                return True
            case _:
                return False

    return len(list(filter(f, Grid(input).cross())))


TEST_INPUT = """MMMSXXMASM
MSAMXMSMSA
AMXSXMAAMM
MSAMASMSMX
XMASAMXAMM
XXAMMXXAMA
SMSMSASXSS
SAXAMASAAA
MAMMMXMMMM
MXMXAXMASX"""

PART1_TESTS = [
    (TEST_INPUT, 18),
]

PART2_TESTS = [
    (TEST_INPUT, 9),
]
